### ISSUE TITLE
We should set a sunset date for Make support in Halide

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#6907]
For several years, we've been maintaining two build systems for Halide (Make and CMake); for a while now, CMake has been the canonical system for building installs, etc., and at least one important feature (Python bindings) aren't support under Make at all anymore. As our build rules have gotten more complex (for things like python-pip, linux installs, etc) it becomes more and more impractical to try to keep Make up to date; this means that when changes inevitably cause a bug in something Make-specific, it often takes an unreasonable amount of time for someone (usually: me) to clean up the issues.

Really, we should just bite the bullet here and do what LLVM did years ago, and commit to CMake as our only build system. As long as we have Make support, someone's gonna be using it; as long as someone is relying on being able to use it, something will break; when something breaks; someone will have to fix it. It's a waste of personpower that our team can't really afford to spend, IMHO.

I'd like to invite anyone who has a stake in using/maintaining the Makefile builds to comment here, but the end goal of this issue is to draw a line in the sand and say "Makefiles will be **removed** from the main Halide branch as of $DATE or $VERSION; CMake will be the only supported build system after that point."

**Comment by abadams** [comment]
I intend to use the Makefile indefinitely. Much of what we do in the research side of the compiler blurs the boundaries between building C++ and managing experiments with dependencies between steps, and cmake only really works adequately for the former. Some of us are going to have to continue writing Makefiles that build on Halide's existing Makefiles indefinitely. 

(deleted rant on why I hate cmake as a C++ build tool, because it's not the actual problem I have)

**Comment by steven-johnson** [comment]
I will gladly meet you in a "Who Hates CMake More" Hate-A-Thon; I usually rate my level at "The fury of a million exploding suns" :-)

But more seriously: every time a nontrivial new build dependency is added, how will it get tested? Right now we're putting a Make Tax on everyone who adds new build dependencies because their PRs won't pass the buildbots without it.

**Comment by abadams** [comment]
A possible incremental approach to avoid losing any make functionality and extensibility while also avoiding repeated build configuration is taking the most stable and complex-to-build parts of the compiler and having the makefile just call out to cmake to build them. I think in practice this means the "distrib" makefile target would ask cmake for a distro. This subset could grow over time, and if cmake wins out ultimately the root makefile will just be syntactic sugar for equivalent cmake command lines.

**Comment by antonysigma** [comment]
> I will gladly meet you in a "Who Hates CMake More" Hate-A-Thon; I usually rate my level at "The fury of a million exploding suns" :-)

I gonna add `Mesonbuild` to the build system multiverse chaos. https://github.com/halide/Halide/issues/6688#issue-1197551640

Meson build language, syntactic similar to Blaze and Android Soong build, aims to provide a more "natural" description language than the Lisp syntax of CMake <=3.09. But I agree that the recent changes in CMake >=3.19 rendered the Meson vs CMake distinction a moot point.

So, in the ticket https://github.com/halide/Halide/issues/6688 , I suggest to register the prebuilt Halide binary on Meson's package system. To avoid having to support yet another build system in this project.

**Comment by zvookin** [comment]
Not clear that the Makefile has to work on anything other than Linux at this point. OS X is nice to have but I don't think it is critical.

**Comment by abadams** [comment]
I would like to keep it working on mac. Many of the students/researchers I'm trying to shield from cmake use mac laptops.

**Comment by zvookin** [comment]
So Linux and Mac OS X. Which functionality needs to work? Obviously building Halide, probably the apps, maybe autoscheduler. Python? Definitely not packaging etc.

I assume the use case is building from source for Halide and probably LLVM, possible some other deps. The CMake instructions are comprehensive, but they seemed to be entirely focused on installing deps as packages, not building from source.

For the MNSU project, we used the Makefiles because cmake did not work. One take away from this is that we probably need to add test coverage for the cmake build instructions. (Though doing so for more than one platform/set of package repositories is likely a nonstarter.)

Basically the issue with cmake is that one has to spool up significant cmake programming skill to work with it and for experimental things one wants to do in builds, it is more effort that it is worth to automate them in cmake to test them out. I also have issues that monoculturing the build system, especially using a really dubious form of specification, is significant long term risk. Like insurance, sometimes one burns a small amount of resources to protect against eventualities in the long term.

It is annoying that we have to repeat a specification of what files are included in what class. Most of the edits to Makefile/CMakeLists.txt could be trivially automated.

**Comment by eli-schwartz** [comment]
> It is annoying that we have to repeat a specification of what files are included in what class. Most of the edits to Makefile/CMakeLists.txt could be trivially automated.

Something I've done for projects that have both meson and cmake is to store the specification of files inside an independent file e.g. `target-sources.lst`, one per line. Then have the build system read that file and split it on `'\n'` to save a variable containing the file list.

This is a bit awkward to do for Make, since you can actually include another Makefile but that requires more structure. You could do `SOURCES = $(shell cat target-sources.lst)` although running subprocesses on every build is slower than just reading a file...



> But I agree that the recent changes in CMake >=3.19 rendered the Meson vs CMake distinction a moot point.

(BTW, I don't think there's any meaningful changes in CMake >=3.19 that can solve the fact that cmake does not have a native array type, but uses POSIX sh style plain strings with magic delimiters, in this case `;` instead of whitespace.)

**Comment by alexreinking** [comment]
> I gonna add Mesonbuild to the build system multiverse chaos.

@antonysigma - sorry, adding yet another build system to Halide is an absolute non-starter and Meson in particular isn't a suitable replacement for Halide's use of CMake. See my closing comment here: https://github.com/halide/Halide/issues/6688#issuecomment-1203113712

> @abadams - Many of the students/researchers I'm trying to shield from cmake use mac laptops.
> @zvookin - For the MNSU project, we used the Makefiles because cmake did not work. [...]

I've been very consistent over the years in my offer to teach CMake to anybody that wants to learn. Yet, nobody has taken me up on it. @zvookin - why not open an issue (did I miss them?) or ask me a question about the CMake issues you and your students observed? @abadams - likewise, you are more than welcome to refer students to me to understand Halide's CMake build.

**Comment by abadams** [comment]
Thanks Alex, though asking students to learn cmake is not something I want them to spend research time on in a limited-duration internship, even with your help. Having two build systems means they can hit the ground running with whichever they're more comfortable with, and they have a backup they can switch to if they get blocked instead of debugging why a build system isn't working on their system, as in Zalman's case.
 
The other half of the issue with research code is that often I'm layering unstructured experiment management rules on top of the Halide build (e.g. adding a build target to an app that generates a paper figure that I'm going to check in to a fork at some weird in-tree relative path, etc), and this sort of ad-hoc nonsense is much simpler in make.

Continuing to maintain the makefiles keeps a low barrier to entry for people who aren't interested in being professional-grade C++ developers.

**Comment by zvookin** [comment]
I believe I discussed the issue on Gitter at the time. The cmake support depended on the LLVM package exporting a cmake support file and the LLVM from ```apt-get install``` did not do that. I think Roman said it was a known issue and was filed on the LLVM side, hence it did not seem appropriate to file an issue on Halide. The suggestion given above is to script the setup instructions and run them on a clean install as a test case, though I doubt we want to do so with high frequency.

(The specific issue of a project's cmake code depending on an installed package exporting cmake support is common. I've run into it multiple times with other stuff I've tried to build. Most recently TaskFlow. Note that one complaint about bazel is it requires this sort of thing. Meson seems to be taking it head on by hosting its own package repository. My guess is the general issue will get resolved for cmake over time simply because everyone will get enough complaints that they'll make sure the repository packages export cmake support, but it'll probably be a while.)

The docs on cmake are different and more comprehensive than they were earlier this year. Yet they're still focused on just building Halide to use Halide, not so much building a lot of stuff from source to do development on the compiler. On the one hand, that there is that much written about cmake setup is probably a good thing. On the other hand, it is way more than anybody wants to read to figure out how to build something just to try it out.

**Comment by alexreinking** [comment]
> @abadams - asking students to learn cmake is not something I want them to spend research time on in a limited-duration internship [...] they have a backup they can switch to if they get blocked instead of debugging why a build system isn't working on their system, as in Zalman's case.

Well, they could still open an issue and assign it to me if something isn't working. I don't think that opening an issue is a huge time investment since I'd be the one debugging. I have to believe you'd have them open an issue (or even a PR) for a Make failure, so why not open an issue for a CMake failure? (In the abstract, I understand Zalman's reason not to in his specific case)

> @zvookin - The docs on cmake are [...] way more than anybody wants to read to figure out how to build something just to try it out.

Yeah, you're absolutely right that the docs on this should be much more straightforward. The main README should probably be condensed to show basic build instructions on up-to-date versions of Windows, macOS, and Ubuntu Linux. I agree with you that those instructions should be tested on CI. That should be easy enough on GitHub Actions. Since I wrote most (all?) of our CMake-related documentation, I'm happy to do all this. I appreciate the feedback. 🙂

**Comment by abadams** [comment]
They're almost always just holding it wrong, or have a messed up system, so when make doesn't work for them I tell them to use cmake. My "have you tried turning it off and on again" is just "have you tried using the other build system".

**Comment by steven-johnson** [comment]
> this sort of ad-hoc nonsense is much simpler in make

well that's both a bug and a feature... it's great for experimenting, but it's not so great when ad-hoc nonsense becomes a long-term build setup, because every single one of us has done this in the past.

**Comment by eli-schwartz** [comment]
> Meson seems to be taking it head on by hosting its own package repository.

Not exactly.

Meson decouples the idea of finding a dependency, and the method used to find it. Whereas in cmake, `find_package() finds cmake scripts and `find_package(PkgConfig)` followed by `pkg_check_modules()` finds pkg-config scripts, meson has a generic function: `dependency('mydep')` and doesn't provide confusing alternative functions.

This dependency can be resolved by various providers, called methods:
- `pkgconfig`: the default, a build-system agnostic dependency export (unlike foo-config.cmake)
- `cmake`: a somewhat hacky thing that tries to see if the dependency can be found by running a dummy cmake project, invoking `find_package()`, and using cmake's trace API to gather details on if it is found and what the results were. This "often" works, is fragile, and requires installing *two* build systems, but if projects insist on only supporting being found via cmake, then it's often helpful
- `config-tool`: some custom stuff meson knows internally about, which is e.g. running sdl2-config, llvm-config, qmake, or other dependencies that provide an executable program taking custom arguments to return Cflags and Libs.
- `system`: libraries that may or may not provide one of the other methods, for example zlib is provided by the platform on some systems such as android, BSD, macOS. This just checks if the compiler can find the libraries
- `builtin`: functionality that may be part of the platform libc, or may be a standalone library. Examples include GNU libintl and GNU libiconv, both of which are baked into GNU libc, and less comprehensive versions can be found in musl libc, the *BSD libc, etc. This will actually return a builtin type dependency for the libc version, or a system type dependency for the standalone library
- `internal`: a dependency that was not otherwise detected, but is provided by an ini file specifying a) the URL to download a subproject buildable with Meson, b) a `dependency_names = foo, bar, baz` list of dependencies that the subproject explicitly exports, or `dep_name = variable_name` mapping of the meson DSL variable that contains a dependency object

Meson's explicit goal is that projects you depend on shouldn't have to care what build system *you* use.

A secondary goal is to help people, especially on Windows, use the "vendor a copy of the dependency and merge it into my own build definitions" style, with a twist: it should be easy to toggle between finding system dependencies and building a vendored copy, and in fact you shouldn't even need to edit your build definitions in order to add a vendored copy, nor check tons of external code into your repo -- that's why ini files are used as "providers".


> resolved for cmake over time simply because everyone will get enough complaints that they'll make sure the repository packages export cmake support

Since it's very hard to do this unless "everyone" using a different build system agrees to switch to cmake (package exports that are written in a turing-complete programming language are not conducive to anything other than the cmake binary internally producing the contents of a Targets file), I think most projects using autotools or meson or waf or scons or bazel or any of a variety of other build systems, will just export a pkg-config file, then tell cmake users to use FindPkgConfig and pkg_check_modules. Because pkg-config files are a very simple key-value format, hence really easy to generate.

**Comment by alexreinking** [comment]
@eli-schwartz - Please refrain from further derailing this issue by disucssing Meson. The place for such discussion would be #6688, but do see the closure reason there.

**Comment by eli-schwartz** [comment]
Nothing I said has anything to do with using meson for halide.

It does have everything to do with how cmake users can look up external dependencies, and export them to other projects too. The same concepts I mentioned regarding provider selection can even be open-coded in a CMakeLists.txt.

**Comment by alexreinking** [comment]
This issue is about whether and to what degree to sunset our Makefiles. I do not see how discussing dependency provision for CMake/Meson adds to this.

**Comment by zvookin** [comment]
Alright folks, by fiat, the issue is now closed. This is going nowhere. Feel free to assign any Makefile issues to @abadams or @zvookin but we will not be sunsetting Makefile support at this time.
